imported/w3c/web-platform-tests/FileAPI/url/url-reload.window.html is timing out in WebKit

cdumez opened the pull request:
#### e295563ee5ae3226673799a8ec66672ed6080473
<pre>
imported/w3c/web-platform-tests/FileAPI/url/url-reload.window.html is timing out in WebKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=244864">https://bugs.webkit.org/show_bug.cgi?id=244864</a>

Reviewed by Geoffrey Garen.

imported/w3c/web-platform-tests/FileAPI/url/url-reload.window.html is timing
out in WebKit but passing in Gecko.

The test is reloading a frame with a Blob URL after revoking that URL,
expecting the reload to succeed. To support this, the Document now keeps its
blob URL alive using a BlobURLHandle.

* LayoutTests/imported/w3c/web-platform-tests/FileAPI/url/url-reload.window-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setURL):
* Source/WebCore/dom/Document.h:

Canonical link: <a href="https://commits.webkit.org/254254@main">https://commits.webkit.org/254254@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/226d984830c7db85a5a5d67313a89d303e18b09f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/88405 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/32864 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/19249 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97587 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/153060 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/92369 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/31362 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26998 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/80603 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92243 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/94012 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24935 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/75269 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24913 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79845 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79951 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67876 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/28973 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/13933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/28963 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14953 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2991 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/32162 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37881 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/31087 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34066 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3500,6 +3500,11 @@ void Document::setURL(const URL& url)
     if (SecurityOrigin::shouldIgnoreHost(m_url))
         m_url.setHostAndPort({ });
 
+    if (m_url.protocolIsBlob())
+        m_blobURLLifetimeExtension = m_url;
+    else
+        m_blobURLLifetimeExtension.clear();
+
```
I wonder if we can / should make a URL container class or subclass, with an internal data member that automatically keeps the Blob URL alive if needed.

Something like:

```
class MightBeBlobURL {
public:
    MightBeBlobURL& operator=(URL& url) {
        m_url = url;
        if (m_url.protocolIsBlob())
            m_blobURLLifetimeExtension = m_url;
    }

    const URL& url() const { return m_url; }

private:
    URL m_url;
    BlobURLHandle m_blobURLLifetimeExtension;
};
```

cdumez replied:
Yes, I think we likely should replace all BlobURLHandle instances with such a class. I can look into this in a follow-up if you don't mind as there are quite a few BlobURLHandle instances.

geoffreygaren replied:
Sure, makes sense.

geoffreygaren reviewed (approved):
r=me

cdumez committed:
imported/w3c/web-platform-tests/FileAPI/url/url-reload.window.html is timing out in WebKit
https://bugs.webkit.org/show_bug.cgi?id=244864

Reviewed by Geoffrey Garen.

imported/w3c/web-platform-tests/FileAPI/url/url-reload.window.html is timing
out in WebKit but passing in Gecko.

The test is reloading a frame with a Blob URL after revoking that URL,
expecting the reload to succeed. To support this, the Document now keeps its
blob URL alive using a BlobURLHandle.

* LayoutTests/imported/w3c/web-platform-tests/FileAPI/url/url-reload.window-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setURL):
* Source/WebCore/dom/Document.h:

Canonical link: https://commits.webkit.org/254254@main
